cooked-validators/src/Cooked/MockChain/Ltl.hs

```diff
@@ -0,0 +1,285 @@
+
+module Cooked.MockChain.Ltl where
+
+import Control.Applicative
+import Control.Monad
+import Control.Monad.State
+
+-- * LTL formulas and operations on them
+
+-- | Type of LTL formulas with atomic formulas of type @a@. Think of @a@ as a
+-- type of "modifications", then a value of type @Ltl a@ describes where to
+-- apply modifications. Since it does not make (obvious) sense to talk of a
+-- negated modification or of one modification (possibly in the future) to
+-- imply another modification, implication and negation are absent.
+data Ltl a
+  = LtlTruth
+  | LtlFalsity
+  | LtlAtom a
+  | -- | Disjunction will be interpreted in an "intuitionistic" way, i.e. as
+    -- branching into the "timeline" where the left disjunct holds and the one
+    -- where the right disjunct holds. In that sense, it is an exclusive or,
+    -- as it does not introduce the branch where both disjuncts hold.
+    LtlOr (Ltl a) (Ltl a)
+  | LtlAnd (Ltl a) (Ltl a)
+  | -- | Assert that the given formula holds at the next time step.
+    LtlNext (Ltl a)
+  | -- | Assert that the first formula holds at least until the second one begins
+    -- to hold, which must happen eventually. The formulas
+    -- > a `LtlUntil` b
+    -- and
+    -- > b `LtlOr` (a `LtlAnd` LtlNext (a `LtlUntil` b))
+    -- are equivalent.
+    LtlUntil (Ltl a) (Ltl a)
+  | -- | Assert that the second formula has to be true up to and including the
+    -- point when the first one becomes true; if that never happens, the second
+    -- formula has to remain true forever. View this as dual to 'LtlUntil'. The
+    -- formulas
+    -- > a `LtlRelease` b
+    -- and
+    -- > b `LtlAnd` (a `LtlOr` LtlNext (a `LtlRelease` b))
+    -- are equivalent.
+    LtlRelease (Ltl a) (Ltl a)
+  deriving (Show)
+
+-- | Split an LTL formula that describes a modification of a computation into a
+-- list of @(doNow, doLater)@ pairs, where
+--
+-- * @doNow@ is the modification to be applied to the current time step,
+--
+-- * @doLater@ is an LTL formula describing the modification that should be
+--   applied from the next time step onwards, and
+--
+-- The return value is a list because a formula might be satisfied in different
+-- ways. For example, the modification described by @a `LtlUntil` b@ might be
+-- accomplished by applying the modification @b@ right now, or by applying @a@
+-- right now and @a `LtlUntil` b@ from the next step onwards; the returned list
+-- will contain these two options.
+--
+-- Modifications should form a monoid, where 'mempty' is the do-nothing
+-- modification, and '<>' is the composition of modifications. Attention: If
+-- '<>' is not commutative, conjunction will also fail to be commutative!
+nowLater :: (Monoid a) => Ltl a -> [(a, Ltl a)]
+nowLater LtlTruth = [(mempty, LtlTruth)]
+nowLater LtlFalsity = []
+nowLater (LtlAtom g) = [(g, LtlTruth)]
+nowLater (a `LtlOr` b) = nowLater a ++ nowLater b
+nowLater (a `LtlAnd` b) =
+  [ (g <> f, ltlSimpl $ c `LtlAnd` d)
```

Comment:
Any particular reason for choosing `g <> f` instead of `f <> g`? In our case, these won't commute